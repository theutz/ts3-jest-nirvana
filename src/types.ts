// Setup a regular structure for the test data.
//
// In this arrangement, the first element of the array is always
// a `string` description of the arguments to be used
// in the `name` parameter of the test.
//
// The arguments are passed as an array, which can be spread into
// the parameters of the system-under-test. They can also be destructured
// to use individually, if desired.
export type TestData<T> = Array<[string, Arguments<T>]>

// Return the type of the arguments as a tuple if T is a function.
export type Arguments<T> = T extends (...args: infer A) => unknown ? A : unknown
