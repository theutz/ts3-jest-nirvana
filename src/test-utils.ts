import { isEmpty } from 'lodash/fp'

// tslint:disable-next-line:ban-types
export function theFunction(fn: Function): string {
  const { name } = fn

  if (isEmpty(name))
    throw new Error('An anonymous function does not have a name.')

  return `${name}()`
}
