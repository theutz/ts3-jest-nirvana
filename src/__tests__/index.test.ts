import { parseStringToNumericArray, stringifySomeNumbers } from '~'
import { theFunction } from '~/test-utils'
import { TestData } from '~/types'

describe(theFunction(parseStringToNumericArray), () => {
  type Data = TestData<typeof parseStringToNumericArray>

  describe('given a string with', () => {
    // This is an example of how you might setup a parameterized test
    // with valid data. If, for whatever reason, the type of the
    // SUT (system-under-test) changes in the source, the TypeScript compiler
    // will complain about the change here.
    //
    // Additionally, it would be trivial to add new test cases, since the
    // form of the test remains the same regardless of the cases presented.
    describe('valid data like', () => {
      const validData: Data = [
        ['undefined', [undefined]],
        ['a number', ['1']],
        ['a series of numbers', ['1,2,-3']]
      ]

      it.each(validData)('%s should return', (_, args) => {
        // This is an example of spreading the arguments into
        // the SUT.
        expect(parseStringToNumericArray(...args)).toMatchSnapshot()
      })
    })

    describe('invalid data like', () => {
      const invalidData: Data = [
        ['a single non-numeric value', ['one']],
        ['a mix of valid and invalid values', ['1,2,three,4']]
      ]

      // This is an example of destructuring the arguments to be used individually.
      it.each(invalidData)('%s should throw', (_, [str]) => {
        expect(() => {
          parseStringToNumericArray(str)
        }).toThrowErrorMatchingSnapshot()
      })
    })
  })
})

// A More Complex Example
//
// The following SUT requires multiple test cases
// to nail down. Luckily, that's SUPER easy with our new setup.
describe(theFunction(stringifySomeNumbers), () => {
  type Data = TestData<typeof stringifySomeNumbers>

  const data: Data = [
    ['a single argument as an array', [[1, 23, 3]]],
    ['an array followed by individual number arguments', [[1], 2, 3, -1048]],
    ['multiple number arguments', [1, 2, 3]],
    ['a single number argument', [1]],
    ['a list of arguments with an array in the middle', [1, [2, -5, 7], 3]],
    ['undefined', [undefined]],
    ['0', [0]]
  ]

  it.each(data)('given %s returns a sorted list', (_, args) => {
    expect(stringifySomeNumbers(...args)).toMatchSnapshot()
  })
})
