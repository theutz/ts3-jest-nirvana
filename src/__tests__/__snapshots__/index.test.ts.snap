// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseStringToNumericArray() given a string with invalid data like a mix of valid and invalid values should throw 1`] = `"'three' is not a number."`;

exports[`parseStringToNumericArray() given a string with invalid data like a single non-numeric value should throw 1`] = `"'one' is not a number."`;

exports[`parseStringToNumericArray() given a string with valid data like a number should return 1`] = `
Array [
  1,
]
`;

exports[`parseStringToNumericArray() given a string with valid data like a series of numbers should return 1`] = `
Array [
  -3,
  1,
  2,
]
`;

exports[`parseStringToNumericArray() given a string with valid data like undefined should return 1`] = `Array []`;

exports[`stringifySomeNumbers() given 0 returns a sorted list 1`] = `"0"`;

exports[`stringifySomeNumbers() given a list of arguments with an array in the middle returns a sorted list 1`] = `"-5,1,2,3,7"`;

exports[`stringifySomeNumbers() given a single argument as an array returns a sorted list 1`] = `"1,3,23"`;

exports[`stringifySomeNumbers() given a single number argument returns a sorted list 1`] = `"1"`;

exports[`stringifySomeNumbers() given an array followed by individual number arguments returns a sorted list 1`] = `"-1048,1,2,3"`;

exports[`stringifySomeNumbers() given multiple number arguments returns a sorted list 1`] = `"1,2,3"`;

exports[`stringifySomeNumbers() given undefined returns a sorted list 1`] = `""`;
